src/librustc/ty/context.rs

```diff
@@ -1097,8 +1103,11 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     }
 
     #[inline(always)]
-    pub fn hir(self) -> &'a hir_map::Map<'gcx> {
-        &self.hir_map
+    pub fn hir(self) -> &'gcx hir_map::Map<'gcx> {
+        self.hir_map.get_or_init(|| {
+            // We can use `with_ignore` here because the hir map does its own tracking
+            self.dep_graph.with_ignore(|| self.hir_map(LOCAL_CRATE))
```

Comment:
We can start to split up the HIR map into several queries, right? Then "its own tracking" might become less and less relevant.

Reply:
I was hoping to split it into per item maps and a whole crate query (for the `Krate` stuff).